src/modules/Game.class.js

```diff
@@ -20,25 +20,297 @@ class Game {
+      // and therefore reverse cache array twice
+
+      cache.reverse();
+
+      for (let i = 0; i < cache.length; i++) {
+        const currentDigit = cache[i];
+        const nextDigit = i + 1 < cache.length ? cache[i + 1] : null;
+
+        if (!nextDigit) {
+          break;
+        }
+
+        if (currentDigit === nextDigit) {
+          const prevScore = this.getScore();
+          const currentScore = prevScore + currentDigit * 2;
+
+          this.score = currentScore;
+
+          cache.splice(i, 2, currentDigit * 2);
+        }
+      }
+
+      cache.reverse();
+
+      // if row doesn't change then go to the next row
+
+      if (cache.length === 4) {
+        continue;
+      }
+
+      // here I need to fill cache array with 0 for equality cache.length === 4
+
+      while (cache.length < 4) {
+        cache.unshift(0);
+      }
+
+      // here I rewrite current row in the state
+
+      for (let cell = 0; cell < 4; cell++) {
+        state[row][cell] = cache[cell];
+      }
+    }
+  }
+
+  moveUp() {
+    const state = this.getState();
+
+    // below I update the state if ArrowUp was pressed
+
+    for (let col = 0; col < 4; col++) {
+      const cache = [];
+
+      // here I push all digits (except 0) of current column in the state
+      // to the cache array
+
+      for (let row = 0; row < 4; row++) {
+        const currentDigit = state[row][col];
+
+        if (!currentDigit) {
+          continue;
+        }
+
+        cache.push(currentDigit);
+      }
+
+      // here I sum equal numbers in the cache array and update the score
+
+      for (let i = 0; i < cache.length; i++) {
+        const currentDigit = cache[i];
+        const nextDigit = i + 1 < cache.length ? cache[i + 1] : null;
+
+        if (!nextDigit) {
+          break;
+        }
+
+        if (currentDigit === nextDigit) {
+          const prevScore = this.getScore();
+          const currentScore = prevScore + currentDigit * 2;
+
+          this.score = currentScore;
+
+          cache.splice(i, 2, currentDigit * 2);
+        }
+      }
+
+      // if column doesn't change then go to the next column
+
+      if (cache.length === 4) {
+        continue;
+      }
+
+      // here I need to fill cache array with 0 for equality cache.length === 4
+
+      while (cache.length < 4) {
+        cache.push(0);
+      }
+
+      // here I rewrite current column in the state
+
+      for (let row = 0; row < 4; row++) {
+        state[row][col] = cache[row];
+      }
+    }
+  }
+
+  moveDown() {
+    const state = this.getState();
+
+    // below I update the state if ArrowDown was pressed
+
+    for (let col = 0; col < 4; col++) {
+      const cache = [];
+
+      // here I push all digits (except 0) of current column in the state
+      // to the cache array
+
+      for (let row = 0; row < 4; row++) {
+        const currentDigit = state[row][col];
+
+        if (!currentDigit) {
+          continue;
+        }
+
+        cache.push(currentDigit);
+      }
```

Comment:
the same with moveDown and moveUp
rotate the columns to rows and call the moveLeft function and then rotate the rows to columns